llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp

```diff
@@ -97,6 +100,95 @@ MDNode *buildSpirvDecorMetadata(LLVMContext &Ctx, uint32_t OpCode,
   return MDNode::get(Ctx, MD);
 }
 
+/// Builds a metadata node for a SPIR-V decoration for cache controls
+/// where decoration code and value are both uint32_t integers.
+/// The value encodes a cache level and a cache control type.
+///
+/// @param Ctx        [in] the LLVM Context.
+/// @param Name       [in] the SPIR-V property string name.
+/// @param OpCode     [in] the SPIR-V opcode.
+/// @param CacheMode  [in] whether read or write.
+/// @param CacheLevel [in] the cache level.
+///
+/// @returns a pointer to the metadata node created for the required decoration
+/// and its values.
+MDNode *buildSpirvDecorCacheProp(LLVMContext &Ctx, StringRef Name,
+                                 uint32_t OpCode, uint32_t CacheMode,
+                                 uint32_t CacheLevel) {
+  // SPIR-V encodings of read control
+  enum cache_control_read_type {
+    read_uncached = 0,
+    read_cached = 1,
+    read_streaming = 2,
+    read_invalidate = 3,
+    read_const_cached = 4
+  };
+  // SPIR-V encodings of write control
+  enum cache_control_write_type {
+    write_uncached = 0,
+    write_through = 1,
+    write_back = 2,
+    write_streaming = 3
+  };
+  // SYCL encodings of read/write control
+  enum cache_mode {
+    uncached,
+    cached,
+    streaming,
+    invalidate,
+    const_cached,
+    through,
+    back
```

Comment:
I'd prefer:
```suggestion
    write_through,
    write_back
```

I think including "write" in the name here would help readability.  These are established terms for how caches work, whereas "through" and "back" by themselves are not.  Including "write" in the name here also makes it clear that passing these two modes through to `read_hint` will give an error.

Reply:
These are not user-visible, but I'll make the change you suggest.